source/Engine.cpp

```diff
@@ -1930,113 +1930,133 @@ void Engine::HandleMouseClicks()
 		
-		// If this projectile has a blast radius, find all ships within its
-		// radius. Otherwise, only one is damaged.
-		double blastRadius = projectile.GetWeapon().BlastRadius();
-		bool isSafe = projectile.GetWeapon().IsSafe();
-		if(blastRadius)
+		// Check if the projectile hit something.
+		if(closestHit < 1.)
 		{
-			// Even friendly ships can be hit by the blast, unless it is a
-			// "safe" weapon.
-			Point hitPos = projectile.Position() + closestHit * projectile.Velocity();
-			for(Body *body : shipCollisions.Circle(hitPos, blastRadius))
+			// Create the explosion the given distance along the projectile's
+			// motion path for this step.
+			projectile.Explode(visuals, closestHit, hitVelocity);
+			
+			// If this projectile has a blast radius, find all ships within its
+			// radius. Otherwise, only one is damaged.
+			double blastRadius = projectile.GetWeapon().BlastRadius();
+			bool isSafe = projectile.GetWeapon().IsSafe();
+			if(blastRadius)
 			{
-				Ship *ship = reinterpret_cast<Ship *>(body);
-				if(isSafe && projectile.Target() != ship && !gov->IsEnemy(ship->GetGovernment()))
-					continue;
-				
-				int eventType = ship->TakeDamage(visuals, projectile.GetWeapon(), 1.,
-					projectile.DistanceTraveled(), projectile.Position(), projectile.GetGovernment(), ship != hit.get());
+				// Even friendly ships can be hit by the blast, unless it is a
+				// "safe" weapon.
+				Point hitPos = projectile.Position() + closestHit * projectile.Velocity();
+				for(Body *body : shipCollisions.Circle(hitPos, blastRadius))
+				{
+					Ship *ship = reinterpret_cast<Ship *>(body);
+					if(isSafe && projectile.Target() != ship && !gov->IsEnemy(ship->GetGovernment()))
+						continue;
+					
+					int eventType = ship->TakeDamage(visuals, projectile.GetWeapon(), 1.,
+						projectile.DistanceTraveled(), projectile.Position(), projectile.GetGovernment(), ship != hit.get());
+					if(eventType)
+						eventQueue.emplace_back(gov, ship->shared_from_this(), eventType);
+				}
+			}
+			else if(hit)
+			{
+				int eventType = hit->TakeDamage(visuals, projectile.GetWeapon(), 1.,
+					projectile.DistanceTraveled(), projectile.Position(), projectile.GetGovernment());
 				if(eventType)
-					eventQueue.emplace_back(gov, ship->shared_from_this(), eventType);
+					eventQueue.emplace_back(gov, hit, eventType);
+			}
+			
+			if(hit)
+			{
+				DoGrudge(hit, gov);
+				hits.insert(hit.get());
 			}
 		}
-		else if(hit)
-		{
-			int eventType = hit->TakeDamage(visuals, projectile.GetWeapon(), 1.,
-				projectile.DistanceTraveled(), projectile.Position(), projectile.GetGovernment());
-			if(eventType)
-				eventQueue.emplace_back(gov, hit, eventType);
-		}
-		
-		if(hit)
-			DoGrudge(hit, gov);
 	}
-	else if(projectile.MissileStrength())
+	
+	// If the projectile is still alive, give the anti-missile systems
+	// a chance to shoot it down.
+	if(projectile.Penetrations() >= 0 && projectile.MissileStrength())
```

Comment:
A `Projectile::IsDead()` function (or some other name) seems appropriate. Also doesn't it make more sense to use `lifetime` instead to check for death?

Reply:
Well a projectile with <0 penetrations or <0 lifetime is dead, so I don't see why one would be more valid than the other if they're both the case for dead projectiles.

Reply:
I guess, but what I though was to not set penetrations to `< 0` at all and to only use `lifetime`.